core/storage/blog_statistics/gae_models.py

```diff
@@ -0,0 +1,863 @@
+# coding: utf-8
+#
+# Copyright 2023 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Models for storing statistics data related to blog posts."""
+
+from __future__ import annotations
+
+from core import utils
+from core.platform import models
+
+from typing import Dict
+
+MYPY = False
+if MYPY: # pragma: no cover
+    from mypy_imports import base_models
+    from mypy_imports import datastore_services
+
+(base_models, ) = models.Registry.import_models([models.Names.BASE_MODEL])
+
+datastore_services = models.Registry.import_datastore_services()
+
+
+class BlogPostViewedEventLogEntryModel(base_models.BaseModel):
+    """An event triggered when a blog post is viewed by any user.
+    The model will be keyed to unique ID which will be of the form -
+    '[timestamp]:[blog_post_id]:[random_hash]'.
+    """
+
+    # ID of blog post currently being viewed.
+    blog_post_id = datastore_services.StringProperty(
+        indexed=True, required=True
+    )
+
+    @staticmethod
+    def get_deletion_policy() -> base_models.DELETION_POLICY:
+        """Model doesn't contain any data directly corresponding to a user."""
+        return base_models.DELETION_POLICY.NOT_APPLICABLE
+
+    @classmethod
+    def get_new_event_entity_id(cls, blog_post_id: str) -> str:
+        """Generates a unique id for the event model of the form
+        '[timestamp]:[blog_post_id]:[random_hash]'.
+        """
+
+        # To avoid collision between events occuring at the same date and time,
+        # a random hash is appended to the event model id.
+        for _ in range(base_models.MAX_RETRIES):
+            random_hash = utils.convert_to_hash(
+                str(utils.get_random_int(base_models.RAND_RANGE)),
+                base_models.ID_LENGTH
+            )
+            new_id = ('%s:%s:%s' % (
+                utils.get_current_time_in_millisecs,
```

Comment:
Do you pass a function here or the return value from the function?

Reply:
And if yes, you should add a test for this behavior because ideally this should have been caught by a test

Reply:
Added. Thanks for pointing this